File: Venda-distribuida-python/TerminalVenda.py
#-*-coding:utf-8 -*-

import getpass
import os
import sys
import time

from proxy.ClientProxy import ClientProxy
from proxy.DatabaseProxy import DatabaseProxy
from proxy.ProdutoProxy import ProdutoProxy
from proxy.VendaProxy import VendaProxy

class TerminalVenda():

    def __init__(self):
        self.clienteSessao = None
        self.cproxy = ClientProxy()
        self.dbproxy = DatabaseProxy()
        self.pproxy = ProdutoProxy()
        self.vproxy = VendaProxy()

    def verificarBase(self):
        return self.dbproxy.verificarBaseReplica()

    def showMain(self):
        print("#######################")
        print("#  1 - Registrar-se   #")
        print("#  2 - Login          #")
        print("#  3 - Sair           #")
        print("#######################")
        option = int(input("Escolha: "))
        if option == 1:
            self.showRegistrar()
        elif option == 2:
            self.showLogin()
        elif option == 3:
            sys.exit()
        else:
            print("Opçao invalida")

    def showRegistrar(self):
        try:
            nome = input("Diga seu nome: ")
            cpf = int(input("Seu CPF (apenas números): "))
            rg = int(input("Seu RG(apenas números): "))
            endereco = input("Onde você mora: ")

            if self.cproxy.registrarCliente(nome,cpf,rg,endereco):
                print("Cliente cadastrado com sucesso")
                self.clienteSessao = {
                    "nome":nome,
                    "cpf":cpf,
                    "rg":rg,
                    "endereco":endereco
                }
                self.showMenu()
            else:
                print("Não foi possível cadastrar o cliente")
                self.showMain()
        except Exception as e:
            print("Informe dados válidos"+str(e))

    def showLogin(self):
        nome = input("Usuario: ")
        cpf = getpass.getpass('CPF (apenas números):')
        self.clienteSessao = self.cproxy.autenticar(nome, cpf)
        if self.clienteSessao is not None:
            os.system('clear')
            print("Bem-vindo! "+self.clienteSessao['nome'])
            while True:
                self.showMenu()
        else:
            self.showMain()

    def showProdutos(self):
        produtos = self.pproxy.listarProdutos()
        for p in produtos:
            print(str(p))

    def realizarCompra(self):
        dataVenda = time.strftime("%Y-%m-%d %H:%M:%S")
        try:
            self.showProdutos()
            produtoId = int(input("Informe o Id do produto que deseja comprar"))
            qtd = int(input("Informe a quantidade de produtos"))
            modoPagamento = int(input("1-Débito\n2-Crédito\nInforme o modo de Pagamento"))
            numCartao = int(input("Informe o número do cartão (apenas número)"))

            prod = None
            for p in self.pproxy.listarProdutos():
                if int(p['id']) == produtoId:
                    prod = p

            if self.vproxy.realizarVenda(self.clienteSessao,prod, qtd, dataVenda,modoPagamento,numCartao):
                print("Compra realizada com sucesso!")
                self.showMenu()
            else:
                print("Não foi possível efetuar a compra!")
                self.showMenu()

        except Exception as e:
            print("Houve um erro ao realizar a compra:"+str(e))

    def showMenu(self):
        os.system('clear')
        print("##########################")
        print("# 1 - Listar Produtos    #")
        print("# 2 - Realizar Compra    #")
        print("# 3 - Sair               #")
        print("##########################")
        option = int(input("Escolha: "))
        if option == 1:
            self.showProdutos()
        elif option == 2:
            self.realizarCompra()
        elif option == 3:
            self.showMain()
        else:
            print("Opção Inválida")


terminal = TerminalVenda()
if(terminal.verificarBase()):
    while True:
        terminal.showMain()
else:
    print("Nao foi possivel fazer a verificaçao da base no mestre!")

File: Venda-distribuida-java/src/main/java/br/ufc/venda/client/client/TerminalVenda.java
package br.ufc.venda.client.client;

import java.util.Calendar;
import java.util.Date;
import java.util.List;

import javax.swing.JOptionPane;

import br.ufc.venda.client.proxy.ClienteProxy;
import br.ufc.venda.client.proxy.DatabaseProxy;
import br.ufc.venda.client.proxy.ProdutoProxy;
import br.ufc.venda.client.proxy.VendaProxy;
import br.ufc.venda.model.Cliente;
import br.ufc.venda.model.ModoPagamento;
import br.ufc.venda.model.Produto;

public class TerminalVenda {

	Cliente clienteSessao;
	ClienteProxy cproxy;
	VendaProxy vproxy;
	ProdutoProxy pproxy;
	DatabaseProxy dbproxy;
	
	public TerminalVenda(){
		cproxy = new ClienteProxy();
		vproxy = new VendaProxy();
		pproxy = new ProdutoProxy();
		dbproxy = new DatabaseProxy();
	}
	
	public boolean verificarBase(){
		return dbproxy.verificarBaseReplica();
	}
	
	private void showMain(){
		int option = 0;
		try{
			option = Integer.parseInt(JOptionPane.showInputDialog("1 - Registrar-se \n2 - Login\n3 - Sair"));
		}catch (Exception e) { 
			JOptionPane.showMessageDialog(null,"Opção inválida!"); 
		}
		if(option == 1)
			showRegister();
		else if(option == 2)
			showLogin();
		else if(option == 3)
			System.exit(0);
		else
			JOptionPane.showMessageDialog(null, "Opção inválida");
	}
	
	private void showRegister(){
		try{
			String nome = JOptionPane.showInputDialog("Diga seu nome:");
			Long cpf = Long.parseLong(JOptionPane.showInputDialog("Diga seu CPF (apenas números):"));
			Long rg = Long.parseLong(JOptionPane.showInputDialog("Diga seu RG (apenas números):"));
			String endereco = JOptionPane.showInputDialog("Diga seu Endereço:");
			Boolean resultado = cproxy.registrarCliente(nome, cpf, rg, endereco);
			System.out.println("TerminalVenda: "+resultado);
			if(resultado){
				JOptionPane.showMessageDialog(null, "Cliente cadastrado com sucesso!", "VENDA-DISTRIBUIDA", JOptionPane.INFORMATION_MESSAGE);
				clienteSessao = new Cliente();
				clienteSessao.setCpf(cpf);
				clienteSessao.setRg(rg);
				clienteSessao.setNome(nome);
				clienteSessao.setEndereco(endereco);
				showMenu();
			}else{
				JOptionPane.showMessageDialog(null, "Não foi possível cadastrar o cliente!", "VENDA-DISTRIBUIDA", JOptionPane.ERROR_MESSAGE);
				showMain();
			}
		}catch (Exception e) { 
			JOptionPane.showMessageDialog(null, "Informe dados válidos: "+e.getMessage()); 
		}
		
	}

	private void showLogin(){
		String usuario = JOptionPane.showInputDialog("Nome:");
		Long senha = Long.parseLong(JOptionPane.showInputDialog("CPF (apenas números):"));
		
		clienteSessao = cproxy.autenticar(usuario, senha);
		if(clienteSessao != null){
			JOptionPane.showMessageDialog(null, "Bem-vindo! "+clienteSessao.getNome(), "VENDA-DISTRIBUÍDA",JOptionPane.INFORMATION_MESSAGE);
			while(true)
				showMenu();
		}else{
			showMain();
		}
	}
	
	private void showProdutos(){
		List<Produto> produtos = pproxy.listarProdutos();
		String lista = "";
		for (Produto produto : produtos) {
			lista += produto+"\n";
		}
		JOptionPane.showMessageDialog(null, lista);
	}
	
	private void realizarCompra(){
		String lista = "";
		Date dataVenda = new Date(Calendar.getInstance().getTimeInMillis());
		try{
			List<Produto> produtos = pproxy.listarProdutos();
			for (Produto produto : produtos) lista += produto+"\n";
			int produtoId = Integer.parseInt(JOptionPane.showInputDialog(null, lista+"Informe o Id do produto que deseja comprar"));
			int qtd = Integer.parseInt(JOptionPane.showInputDialog(null, "Informe a quantidade de produtos"));
			int modoPagamento = Integer.parseInt(JOptionPane.showInputDialog(null, "1-Débito\n2-Crédito\nInforme o modo de Pagamento"));
			Long numCartao = Long.parseLong(JOptionPane.showInputDialog(null, "Informe o número do cartão (apenas número)"));
			
			Produto prod = null;
			for (Produto produto : produtos) if(produto.getId() == produtoId) prod = produto;
			if(modoPagamento != ModoPagamento.CREDITO.pagamento() && modoPagamento != ModoPagamento.DEBITO.pagamento())
				return;
			
			Boolean resultado = vproxy.realizarVenda(clienteSessao, prod, qtd, dataVenda, modoPagamento, numCartao);
			if(resultado){
				JOptionPane.showMessageDialog(null, "Compra realizada com sucesso!", "VENDA-DISTRIBUIDA", JOptionPane.INFORMATION_MESSAGE);
				showMenu();
			}else{
				JOptionPane.showMessageDialog(null, "Não foi possível efetuar a compra!", "VENDA-DISTRIBUIDA", JOptionPane.ERROR_MESSAGE);
				showMenu();
			}
				
		}catch (Exception e) {
			JOptionPane.showMessageDialog(null, "Houve um erro ao realizar a compra: "+e.getMessage());
		}
	}
	
	private void showMenu(){
		int option = 0;
		try{
			option = Integer.parseInt(JOptionPane.showInputDialog("1 - Listar Produtos \n2 - Realizar Compra\n3 - Sair"));
		}catch (Exception e) {
			JOptionPane.showMessageDialog(null, "Opção inválida");
		}
		if(option == 1)
			showProdutos();
		else if(option == 2)
			realizarCompra();
		else if(option == 3)
			showMain();
		else
			JOptionPane.showMessageDialog(null, "Opção inválida");
	}
	
	public static void main(String[] args) {
		TerminalVenda tv = new TerminalVenda();
		if(tv.verificarBase())
			while(true)
				tv.showMain();
		else
			JOptionPane.showMessageDialog(null, "A base deste cliente não está no mestre!", "VENDA-DISTRIBUÍDA", JOptionPane.ERROR_MESSAGE);
	}

}
